Sources/Workspace/Workspace.swift

```diff
@@ -1326,25 +1326,86 @@ extension Workspace {
 // MARK: - Pinning Functions
 
 extension Workspace {
-
     /// Pins all of the current managed dependencies at their checkout state.
-    fileprivate func pinAll(
+    fileprivate func saveResolvedFile(
         pinsStore: PinsStore,
         dependencyManifests: DependencyManifests,
         rootManifestsMinimumToolsVersion: ToolsVersion,
         observabilityScope: ObservabilityScope
-    ) {
+    )  {
+        var dependenciesToPin = [ManagedDependency]()
+        let requiredDependencies = dependencyManifests.computePackages().required.filter({ $0.kind.isPinnable })
+        for dependency in requiredDependencies {
+            if let managedDependency = self.state.dependencies[comparingLocation: dependency] {
+                dependenciesToPin.append(managedDependency)
+            } else {
+                observabilityScope.emit(warning: "required dependency \(dependency.identity) (\(dependency.locationString)) was not found in managed dependencies and will not be recorded in resolved file")
+            }
+        }
+
+        // FIXME
+        // load the pin store from disk so we can compare for any changes
+        // this is needed as we want to avoid re-writing the resolved files
+        // unless absolutely required
+        guard let storedPinStore = observabilityScope.trap({ try self.pinsStore.load() }) else {
```

Comment:
So if this fails, such as because of an I/O error, would it be better to try to write the new file anyway?  I guess I'm questioning whether returning here is the right thing.

Reply:
are you suggesting we try rewrite if we cant read? I can go with that too

Reply:
Yes, exactly (I should have been more clear).  Since the only reason we're reading it is to avoid unnecessary changes, it seems reasonable to write it out if the read fails.  Of course, if it's a permission issue, the write may still fail, but that's a separate issue.

Reply:
updated https://github.com/apple/swift-package-manager/pull/3967/commits/e2375162f06a66a1e0cd91b11aecd67e2520d8e0